doc/PLUGINS.md

```diff
@@ -547,6 +547,44 @@ there's a member `error_message`, that member is sent to the peer
 before disconnection.
 
 
+### `commitment_revocation`
+
+State updates in Lightning consist of the following steps:
+
+ 1. Proposal of a new state commitment in the form of a commitment transaction
+ 2. Exchange of signatures for the agreed upon commitment transaction
+ 3. Verification that the signatures match the commitment transaction
+ 4. Exchange of revocation secrets that could be used to penalize an eventual misbehaving party
+
+The `commitment_revocation` hook is used to inform the plugin about the state
+transition being completed, and deliver the penalty transaction. The penalty
+transaction could then be sent to a watchtower that automaticaly reacts in
+case one party attempts to settle using a revoked commitment.
+
+The payload consists of the following information:
+
+```json
+{
+	"commitment_txid": "58eea2cf538cfed79f4d6b809b920b40bb6b35962c4bb4cc81f5550a7728ab05",
+	"penalty_tx": "02000000000101...ac00000000"
+}
+```
+
+Notice that the `commitment_txid` could also be extracted from the sole input
+of the `penalty_tx`, however it is enclosed so plugins don't have to include
+the logic to parse transactions.
+
+Not included are the `htlc_success` and `htlc_failure` transactions that
+may also be spending `commitment_tx` outputs. This is because these
+transactions are much more dynamic and have a predictable timeout, allowing
+wallets to ensure a quick checkin when the CLTV of the HTLC is about to
+expire.
+
+The `commitment_revocation` hook is a chained hook, i.e., multiple plugins can
+register it, and they will be called in the order they were registered in.
+Plugins should always return `{"result": "continue"}`, otherwise subsequent
+hook subscribers would not get called.
```

Comment:
What would the use-case be for interactivity of hooks?  Is there any reason for a plugin to somehow affect state update?  Would not `htlc_accepted` (and maybe a future `ptlc_accepted`) be sufficient as a hook, since we will largely be manipulating HTLC/PTLCs in the future only?  It seems to me more sensible to use a notification for `commitment_revocation` event, and use a hook for `htlc_accepted`.

We can only emit this event (whether notification or hook) *after* the counterparty sends a `revoke_and_ack` to us.  This is pretty much the last step in the update dance (I think we send our own `revoke_and_ack` even before we know of the counterparty `revoke_and_ack`, as otherwise it would be a chicken-and-egg problem), so I do not see *what* a hook can affect for this state: it would be "too late" at this point to do anything about this state, so what can a plugin affect for this state?  Even if we disconnect at this point, as far as the counterparty is concerned the new state is now the latest (and it will re-send the revocation data on channel re-establish).

Reply:
Hmm.

What it *could* affect would be dependent processing: i.e. we do not consider any HTLCs introduced by the new state to be "irrevoably committed" until the most recent `commitment_revocation` hook returns.  This implies adding a new state in the database: we know the revocation of a previous state, but have not yet confirmed with `commitment_revocation` that all the plugins have confirmed they have processed this revocation correctly, in which case on a restart we will re-invoke `commitment_revocation` on all plugins.  It seems safer to implement this first if we are going to make this a hook anyway.

Reply:
So, in case of a failure in `commitment_revocation` (i.e. it returns anything other than a non-error result `{"result": "continue"}`), then what we should do is to store the failed `commitment_revocation` on-database, then set the channel to "always fail" mode for now.

Then future updates on the channel, we:

* We receive `revoke_and_ack`.
* Append the new `commitment_revocation` to the list in he database.
* For every `commitment_revocation` in the in-database list (in order of appending!):
  * Send `commitment_revocation` to the plugin hook chain.
  * If it succeeds, remove that from the in-database list and continue with the next item in the list.
  * If it fails, exit this loop and mark all live incoming HTLCs as failing with `temporary_node_failure`.
* If the above loop completely succeeded, the in-database list would be empty.
* If the above loop completely succeeded, then we can safely handle (forward or claim) any incoming HTLCs.

On startup we perform the above loop as well, and handle incoming HTLCs as normal if the loop completes.

This allows us to keep the node "alive" in some way if its watchtower dies.

Reply:
> What would the use-case be for interactivity of hooks? Is there any reason for a plugin to somehow affect state update? Would not `htlc_accepted` (and maybe a future `ptlc_accepted`) be sufficient as a hook, since we will largely be manipulating HTLC/PTLCs in the future only? It seems to me more sensible to use a notification for `commitment_revocation` event, and use a hook for `htlc_accepted`.

The reason I like it being a hook is that if we decide that notifying watchtowers is as critical as it's assumed to be, we can make sure that the watchtower gets the `penalty_tx` before we make progress. This can be achieved by splitting the `peer_got_revoke` function, then sending the `towire_channel_got_revoke_reply` and save to DB only after the watchtower plugin gave us the all-clear. If the watchtower is currently unreachable it is perfectly fine for a plugin to return a `{"result": "fail"}` signalling that we must abort since proceeding could mean that our counterparty publishes this specific state when we're offline, and the watchtower cannot save us. Aborting and restarting means we get another chance later, since our peer will resend the last round of commitment and revocation, at which point the watchtower hopefully is operational again.

I'm not sure myself whether the watchtower hook is that important, but it's nice to have the option :-) 'm not proposing any other state modification, since those are better suited for `htlc_accepted` and `ptlc_accepted` as you correctly point out.

Reply:
There's no reason for an intermediate state.  If we receive the revocation and don't complete all the hooks, we will ask the peer for it again, just as if we'd never received it.